lib/proforma.rb

```diff
@@ -14,7 +15,17 @@
 require 'proforma/models/task'
 
 module Proforma
-  SCHEMA_FORMAT_PATH = File.join(File.dirname(File.expand_path(__FILE__)), '../assets/schemas/proforma_%s.xsd')
-  SCHEMA_VERSIONS = ['2.1', '2.0'].freeze
+  SCHEMA_PATH = File.join(File.dirname(File.expand_path(__FILE__)), '../assets/schemas')
+  SCHEMA_FORMAT_PATH = "#{SCHEMA_PATH}/proforma-%s.xsd".freeze
+  SCHEMA_VERSIONS = %w[2.1 2.0].freeze
+
+  TEST_TYPE_SCHEMA_NAMES = %w[checkstyle regexptest unittest].freeze
+  TEST_TYPE_SCHEMAS = {}.tap do |hash|
```

Comment:
I might be missing something, but are you actually using these test type schemas, e.g., to validate the input?

Reply:
You are correct, I'm not using the schemas at all. I had difficulties to include them in the validationprocess and opted for just allowing and saving anything it that tag.

Reply:
Mh, some validation would be nice, of course. Can you assist me in understanding these difficulties? Is it because you cannot select the appropriate XSD or because validating the partial just doesn't work?

Reply:
- yes validating the partial didn't work
- a more general question is: when should we validate. In the proforma.xsd the node in `test-configuration` is just an any-node without any reference to the other xsds. Only [the proforma whitepaper](https://github.com/ProFormA/proformaxml/blob/master/Whitepaper.md#appendix-c-list-of-test-types) describes that there are some `test-types` that have xsds. We could stick to that and hardcode those `test-types`.

Reply:
Thanks for outlining and trying again 👍 

Regarding your question, I would go with a validation if possible and indicated by a (known) schema. So, if the `test-configuration` contains nodes with another namespace, all those nodes with the other namespace should be validated, if the namespace belongs to those prefixed with `urn:proforma:tests`. In this case, all nodes should be valid with regard to the XSD and rejected otherwise. Does this make sense to you?

Reply:
Yes, I think I understand the idea.

I found a way to validate the partial, but I think it's very inelegant (it involves using Nokogiri to export a node  to string and parse it again, multiple times) - but at least it seems to work

Reply:
Awesome, thank you! Based on your current implementation, I refactored that a little and removed the string parsing. What do you think? I would argue that this solution is cleaner and more elegant.

Unrelated to this change, we should also remove those unneeded references in `lib/proforma.rb` if no longer needed.

<details><summary>Suggested Code Change</summary>
<p>

```diff
Subject: [PATCH] Refactor test config validation
---
Index: lib/proforma/services/validator.rb
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/proforma/services/validator.rb b/lib/proforma/services/validator.rb
--- a/lib/proforma/services/validator.rb	(revision c3ccd5426e176873d234a70f377e3ba3778cf4a6)
+++ b/lib/proforma/services/validator.rb	(date 1689611208641)
@@ -17,42 +17,36 @@
       @doc_schema_version ||= /^urn:proforma:v(.*)$/.match(@doc.namespaces['xmlns'])&.captures&.dig(0)
     end
 
-    def custom_test_config_validation
-      return [] if @doc.namespaces.values.none? {|ns| ns.start_with?('urn:proforma:tests:') }
-
-      validate_proforma_test_configuration
-    end
-
-    def validate_proforma_test_configuration
-      @doc.xpath('/xmlns:task/xmlns:tests/xmlns:test/xmlns:test-configuration').map do |test_config|
-        test_config.children.select do |c|
-          c.namespace&.href&.start_with?('urn:proforma:tests:')
-        end.map {|config| validate_proforma_config_node config }
-      end.flatten
-    end
-
-    def validate_proforma_config_node(config_node)
-      tmp_doc = node_as_doc_with_namespace(config_node)
-
-      return unless TEST_TYPE_SCHEMA_NAMES.include? config_node.name
-
-      schema = Nokogiri::XML::Schema(File.read(Proforma::SCHEMA_FORMAT_PATH % config_node.name))
-      schema.validate(Nokogiri::XML(tmp_doc.to_xml))
-    end
-
     def node_as_doc_with_namespace(config_node)
-      Nokogiri::XML(config_node.to_xml).tap do |doc|
-        doc.children.first.add_namespace_definition(config_node.namespace.prefix, config_node.namespace.href)
-      end
+      doc = Nokogiri::XML::Document.new
+      doc.add_child(config_node)
+      doc
     end
 
     def validate
+      # Both validations return an array of errors, which are empty if the validation was successful.
+      validate_task + validate_test_configuration
+    end
+
+    def validate_task
       return ['no proforma version found'] if doc_schema_version.nil?
 
       version = @expected_version || doc_schema_version
       return ['version not supported'] unless SCHEMA_VERSIONS.include? version
 
-      Nokogiri::XML::Schema(File.open(SCHEMA_FORMAT_PATH % version)).validate(@doc) + custom_test_config_validation
+      Nokogiri::XML::Schema(File.open(SCHEMA_FORMAT_PATH % version)).validate(@doc)
+    end
+
+    def validate_test_configuration
+      @doc.xpath('/xmlns:task/xmlns:tests/xmlns:test/xmlns:test-configuration').flat_map do |test_config|
+        test_config.children.flat_map do |config_node|
+          next [] unless config_node.namespace&.href&.start_with?('urn:proforma:tests:')
+          next [] unless TEST_TYPE_SCHEMA_NAMES.include? config_node.name
+
+          schema = Nokogiri::XML::Schema(File.read(SCHEMA_FORMAT_PATH % config_node.name))
+          schema.validate(node_as_doc_with_namespace(config_node))
+        end
+      end
     end
   end
 end
```

</p>
</details>

Reply:
Nice approach, thanks!

There were however two small problems:
1. the validation was now trying to run before the check, if the xml is actually trying to be a proforma (validator.rb line 27 & 30) - I pulled those checks to the front again
2. You used `doc.add_child(config_node)`. Turns out the "reparents" the node, so it was missing from the original document. (Fixed by adding a dup of the node)

Reply:
Happy you were able to build on my suggestion. Even though I was testing the code using an example and my debugger, I obviously missed the "reparenting" of the `config_node` and the general check of the ProFormA format. Probably, it would have been a good idea to run the tests :D

Nevertheless, I think it was good team work to get this done ;).